src/shared/bus-util.h

```diff
@@ -23,11 +23,31 @@ typedef enum BusTransport {
 } BusTransport;
 
 typedef struct BusLocator {
-        const char    *destination;
-        const char    *path;
-        const char    *interface;
+        const char *destination;
+        const char *path;
+        const char *interface;
 } BusLocator;
 
+typedef struct BusObjectImplementation BusObjectImplementation;
+
+typedef struct BusObjectVtablePair {
+        const sd_bus_vtable *vtable;
+        sd_bus_object_find_t object_find;
+} BusObjectVtablePair;
+
+struct BusObjectImplementation {
+        const char *path;
+        const char *interface;
+        const BusObjectVtablePair *vtables;
+        sd_bus_node_enumerator_t node_enumerator;
+        bool manager;
+        const BusObjectImplementation* children[];
+};
+
+#define VTABLE_LIST(...) ((const BusObjectVtablePair[]) { __VA_ARGS__, {} })
+#define CHILD_LIST(...) { __VA_ARGS__ __VA_OPT__(,) NULL }
```

Comment:
hmm, one more thing: is there really such a big benefit of listing them children inline of the structure, while listing the vtables via an external array? 

Having one or multiple vtables is going to be the common case, while having children is the exception, hence if anything this should be reversed, no? i.e. that the vtables are listed in-line while the children are listed as array?

that said, I think it would be a lot prettier if we'd do both arrays as external arrays, no? so that the structure becomes fixed size, and we don't have to deal with zero-initialization of that field?

Reply:
That's a good point. I'll rework it like that.